crates/telio-traversal/src/connectivity_check/cross_ping_check.rs

```diff
@@ -577,6 +602,61 @@ impl<E: Backoff> State<E> {
 
         Ok(())
     }
+
+    pub fn session_for_key(&self, pk: PublicKey) -> Option<Session> {
+        self.endpoint_connectivity_check_state
+            .iter()
+            .find(|(_, state)| state.public_key == pk)
+            .map(|(session, _)| *session)
+    }
+
+    pub fn check_if_upgrade_is_allowed(&mut self, session: Session, public_key: PublicKey) -> bool {
+        match self.session_id_candidates.remove(&session) {
```

Comment:
Generally the lifetime of the session in this case is tied to endpoint candidate and peer being in meshnet config. Where sessions are instantiated [here](https://github.com/NordSecurity/libtelio/blob/main/crates/telio-traversal/src/connectivity_check/cross_ping_check.rs#L373) and [here](https://github.com/NordSecurity/libtelio/blob/main/crates/telio-traversal/src/connectivity_check/cross_ping_check.rs#L434) respectively when peer list has changed or when endpoint (local) list has changed. 

Therefore it is valid to have multiple upgrade sessions for the same session, thus maybe we should not remove session id from candidates when upgrade is received?

Reply:
True, but I think that we other side will always get that session id in a call me maybe, so when we check if it allowed it will already be there even if we remove it while checking it.

Reply:
Oh I was thinking about the case where single CPC cycle is used for multiple upgrades.

Reply:
Ok, I double-checked a bit, and now we are emitting "endpoint gone" event on every "downgrade" which would mean, that probably there are no cases where single CPC cycle is used for multiple upgrades. Just not sure if we want to couple these two things together.

Reply:
You know, after realizing "endpoint gone" thing and thinking a bit more about this -> I have changed my mind, I think your suggestion is better :P